stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,829 @@
+  std::vector<arena_t<T_ts>, arena_allocator<arena_t<T_ts>>> ts_;
+
+  double relative_tolerance_forward_;
+  arena_t<Eigen::VectorXd> absolute_tolerance_forward_;
+  double relative_tolerance_backward_;
+  arena_t<Eigen::VectorXd> absolute_tolerance_backward_;
+  double relative_tolerance_quadrature_;
+  double absolute_tolerance_quadrature_;
+  long int max_num_steps_;                  // NOLINT(runtime/int)
+  long int num_steps_between_checkpoints_;  // NOLINT(runtime/int)
+  int interpolation_polynomial_;
+  int solver_forward_;
+  int solver_backward_;
+
+  std::ostream* msgs_;
+
+  std::tuple<arena_t<
+      plain_type_t<decltype(deep_copy_vars(std::declval<const T_Args&>()))>>...>
+      local_args_tuple_;
+  std::tuple<arena_t<
+      plain_type_t<decltype(value_of(std::declval<const T_Args&>()))>>...>
+      value_of_args_tuple_;
+
+  size_t N_;
+  bool forward_returned_;
+  bool backward_is_initialized_;
+
+  size_t num_t0_vars_;
+  size_t num_ts_vars_;
+  size_t num_y0_vars_;
+  size_t num_args_vars_;
+  size_t num_vars_;
+
+  arena_t<Eigen::VectorXd> state_forward_;
+  arena_t<Eigen::VectorXd> state_backward_;
+  arena_t<Eigen::VectorXd> quad_;
+
+  vari** non_chaining_varis_;
+
+  vari** t0_varis_;
+  vari** ts_varis_;
+  vari** y0_varis_;
+  vari** args_varis_;
+
+  int index_backward_;
+
+  struct cvodes_solver;
+  cvodes_solver* solver_;
+
+  static constexpr bool is_var_ts{is_var<T_ts>::value};
+  static constexpr bool is_var_t0{is_var<T_t0>::value};
+  static constexpr bool is_var_y0{is_var<T_y0_t0>::value};
+  static constexpr bool is_any_var_args{
+      disjunction<is_var<scalar_type_t<T_Args>>...>::value};
+  static constexpr bool is_var_return{is_var<T_Return>::value};
+  static constexpr bool is_var_only_ts{
+      is_var_ts && !(is_var_t0 || is_var_y0 || is_any_var_args)};
+
+  /**
+   * Call the ODE RHS with given tuple.
+   */
+  template <typename yT, typename... ArgsT>
+  constexpr auto rhs(double t, const yT& y,
+                     const std::tuple<ArgsT...>& args_tuple) const {
+    return apply([&](auto&&... args) { return f_(t, y, msgs_, args...); },
+                 args_tuple);
+  }
+
+  /**
+   * Implements the function of type CVRhsFn which is the user-defined
+   * ODE RHS passed to CVODES.
+   */
+  constexpr static int cv_rhs(realtype t, N_Vector y, N_Vector ydot,
+                              void* user_data) {
+    const cvodes_integrator_adjoint_vari* integrator
+        = static_cast<const cvodes_integrator_adjoint_vari*>(user_data);
+    integrator->rhs(t, NV_DATA_S(y), NV_DATA_S(ydot));
+    return 0;
+  }
+
+  /**
+   * Implements the function of type CVRhsFnB which is the
+   * RHS of the backward ODE system.
+   */
+  constexpr static int cv_rhs_adj_sens(realtype t, N_Vector y, N_Vector yB,
+                                       N_Vector yBdot, void* user_data) {
+    const cvodes_integrator_adjoint_vari* integrator
+        = static_cast<const cvodes_integrator_adjoint_vari*>(user_data);
```

Comment:
idt these should be const, it's usually odd to have a const vari or vari pointer because the stuff inside of the vari is certainly not const

Reply:
why not declare the object const if we know that it won't be changed? I thought this helps the compiler to optimize, no?

Reply:
const went away